xbout/geometries.py

```diff
@@ -180,19 +180,34 @@ def apply_geometry(ds, geometry_name, *, coordinates=None, grid=None):
         # Generates a coordinate whose value is 0 on the first grid point, not dz/2, to
         # match how BOUT++ generates fields from input file expressions.
         nz = updated_ds.dims[zcoord]
-        z0 = 2 * np.pi * updated_ds.metadata["ZMIN"]
-        z1 = z0 + nz * updated_ds.metadata["dz"]
-        if not np.isclose(
-            z1, 2.0 * np.pi * updated_ds.metadata["ZMAX"], rtol=1.0e-15, atol=0.0
-        ):
-            warn(
-                f"Size of toroidal domain as calculated from nz*dz ({str(z1 - z0)}"
-                f" is not the same as 2pi*(ZMAX - ZMIN) ("
-                f"{2.*np.pi*updated_ds.metadata['ZMAX'] - z0}): using value from dz"
+
+        # In BOUT++ v5, dz is either a Field2D or Field3D.
+        # We can use it as a 1D coordinate if it's a Field3D, _or_ if nz == 1
+        bout_v5 = updated_ds.metadata["BOUT_VERSION"] >= 5.0
+        use_metric_3d = updated_ds.metadata.get("use_metric_3d", False)
+        can_use_1d_z_coord = (nz == 1) or use_metric_3d
+
+        if can_use_1d_z_coord:
+            z = _1d_coord_from_spacing(updated_ds["dz"], zcoord)
+        else:
+            if bout_v5:
+                dz = updated_ds["dz"][0, 0]
```

Comment:
It would be nice to check that `dz` is actually constant here. I think at the moment it always should be, but you never know what someone will want to do in future...
```suggestion
                if not np.all(updated_ds["dz"].min() == updated_ds["dz"].max()):
                    raise ValueError(
                        f"Spacing is not constant. Cannot create z coordinate"
                    )
                dz = updated_ds["dz"][0, 0]
```

Reply:
Good shout :+1: 

xarray should presumably be able to support coordinates that vary across multiple dimensions, so at some point we should be able to set this "properly". This would of course be much easier if BOUT++ wrote out the proper coordinates and not just the spacings and metric!

Reply:
`xarray` can indeed support multidimensional coordinates (we load for example `Rxy` and `Zxy` as 2d (or now 3d I guess) `R` and `Z` coordinates).

The wrinkle is that it's useful also to have a 1d coordinate associated with each dimension - what `xarray` calls a 'dimension coordinate'. The reason it's useful to have those dimension coordinates defined is that it means that when you do a selection (for example) you don't lose track of where in a dimension the data is defined. `xarray`'s broadcasting machinery can use these 'dimension coordinates' to do clever things (see e.g. http://xarray.pydata.org/en/stable/user-guide/computation.html?highlight=alignment#automatic-alignment).

Especially with 3d coordinates coming in, it might well make sense to assume our `x`, `y`, `z` coordinates may be multidimensional, and just attach a global index 'coordinate', e.g. (loosely) `ds["xind"] = GlobalXIndex()`, `ds["yind"] = GlobalYIndex()`, `ds["zind"] = GlobalZIndex()`. I've thought about doing this before, and just not had a strong enough reason when the regular `y` and `z` coordinates were good 1d coordinates. `x` is already doing this, because `psi` is in general 2d, so BOUT++'s 'x-coordinate' isn't a 1d coordinate (it can be defined differently between PFR and core, or between inner and outer SOL in a double-null), so in `xbout` `x` is just the global index number, and `psi_poloidal` (if I remember the name right) is a 2d coordinate. That'd be a change for a separate PR though!

Reply:
Yes, definitely should be another PR!

I'm struggling a bit to get my head around what's the correct thing to do here. Is the bit below, checking that `nz * dz == 2pi*(ZMAX - ZMIN)`\* strictly necessary? If not, then I think this whole bit could boil down to:

1. Move `dz` from `updated_ds.metadata` to `updated_ds`
2. Use `_1d_coord_from_spacing`, with the proviso that this should actually add `ZMIN` instead of `0.5` to the coordinate
3. Done?

---
\* The correct solution is probably that BOUT++ should error if `ZMIN, ZMAX, dz` are all set

Reply:
I think the `+ 0.5` in `_1d_coord_from_spacing()` should stay - at least the one I'm looking at is taking into account the offset between the first cell edge (where the coordinate 'starts') and cell centre (where coordinate is evaluated). So like you say, use `_1d_coord_from_spacing()`, but then just add `ZMIN`.

The check with `ZMAX` isn't strictly necessary, but I think I added it because I got bitten by an input file which defined both `dz` and `ZMAX`, but inconsistently, so I'd like to keep the check in.

> * The correct solution is probably that BOUT++ should error if ZMIN, ZMAX, dz are all set

Agreed :+1:

Reply:
If I add the check for a constant `dz`
```suggestion
                if not np.all(updated_ds["dz"].min() == updated_ds["dz"].max()):
                    raise ValueError(
                        f"Spacing is not constant. Cannot create z coordinate"
                    )
                dz = updated_ds["dz"][0, 0]
```
will it cause problems (e.g. 3D `dz` that's 'constant' but only up to machine precision)?

Reply:
Could use `allclose` instead to account for that?

Reply:
I think `==` is fine - we use the same in BOUT++ (`isUniform`):
https://github.com/boutproject/BOUT-dev/blob/next/include/field.hxx#L369